server/etcdserver/api/v3discovery/discovery_test.go

```diff
@@ -0,0 +1,651 @@
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	"go.etcd.io/etcd/client/pkg/v3/types"
+	"go.etcd.io/etcd/client/v3"
+
+	"github.com/jonboulle/clockwork"
+	"go.uber.org/zap"
+)
+
+var (
+	clusterFakeToken = "fakeToken"
+
+	actualMemberInfo = []memberInfo{
+		{
+			peerRegKey:  "/_etcd/registry/fakeToken/members/" + types.ID(101).String(),
+			peerURLsMap: "infra1=http://192.168.0.100:2380",
+			createRev:   8,
+		},
+		{
+			// invalid peer registry key
+			peerRegKey:  "/invalidPrefix/fakeToken/members/" + types.ID(102).String(),
+			peerURLsMap: "infra2=http://192.168.0.102:2380",
+			createRev:   6,
+		},
+		{
+			peerRegKey:  "/_etcd/registry/fakeToken/members/" + types.ID(102).String(),
+			peerURLsMap: "infra2=http://192.168.0.102:2380",
+			createRev:   6,
+		},
+		{
+			// invalid peer info format
+			peerRegKey:  "/_etcd/registry/fakeToken/members/" + types.ID(102).String(),
+			peerURLsMap: "http://192.168.0.102:2380",
+			createRev:   6,
+		},
+		{
+			peerRegKey:  "/_etcd/registry/fakeToken/members/" + types.ID(103).String(),
+			peerURLsMap: "infra3=http://192.168.0.103:2380",
+			createRev:   7,
+		},
+		{
+			// duplicate peer
+			peerRegKey:  "/_etcd/registry/fakeToken/members/" + types.ID(101).String(),
+			peerURLsMap: "infra1=http://192.168.0.100:2380",
+			createRev:   2,
+		},
+	}
+
+	// sort by CreateRevision
+	expectedMemberInfo = []memberInfo{
+		{
+			peerRegKey:  "/_etcd/registry/fakeToken/members/" + types.ID(102).String(),
+			peerURLsMap: "infra2=http://192.168.0.102:2380",
+			createRev:   6,
+		},
+		{
+			peerRegKey:  "/_etcd/registry/fakeToken/members/" + types.ID(103).String(),
+			peerURLsMap: "infra3=http://192.168.0.103:2380",
+			createRev:   7,
+		},
+		{
+			peerRegKey:  "/_etcd/registry/fakeToken/members/" + types.ID(101).String(),
+			peerURLsMap: "infra1=http://192.168.0.100:2380",
+			createRev:   8,
+		},
+	}
+)
+
+// fakeKVForClusterSize is used to test getClusterSize.
+type fakeKVForClusterSize struct {
+	*fakeBaseKV
+	clusterSizeStr string
+}
+
+// We only need to overwrite the method `Get`.
+func (fkv *fakeKVForClusterSize) Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	if fkv.clusterSizeStr == "" {
+		// cluster size isn't configured in this case
+		return &clientv3.GetResponse{}, nil
+	}
+
+	return &clientv3.GetResponse{
+		Kvs: []*mvccpb.KeyValue{
+			{
+				Value: []byte(fkv.clusterSizeStr),
+			},
+		},
+	}, nil
+}
+
+func TestGetClusterSize(t *testing.T) {
+	cases := []struct {
+		clusterSizeStr string
+		expectedErr    error
+		expectedSize   int
+	}{
+		{
+			clusterSizeStr: "",
+			expectedErr:    ErrSizeNotFound,
+		},
+		{
+			clusterSizeStr: "invalidSize",
+			expectedErr:    ErrBadSizeKey,
+		},
+		{
+			clusterSizeStr: "3",
+			expectedErr:    nil,
+			expectedSize:   3,
+		},
+	}
+
+	lg, err := zap.NewProduction()
+	if err != nil {
+		t.Errorf("Failed to create a logger, error: %v", err)
+	}
+	for i, tc := range cases {
```

Comment:
Please use named subtests (explained in other comment).

Reply:
Resolved & thanks.